rfcs/45-stageless.md

```diff
@@ -0,0 +1,584 @@
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them
+   // `SystemId` should be a unique opaque identifier, generated on system insertion
+   // We cannot use a Vec + a `usize` index as the identifier,
+   // as the identifier would not be stable when systems were removed, risking serious bugs
+   systems: HashMap<SystemId, ConfiguredSystem>,
+   // Label configuration must be stored at the schedule level,
+   // rather than attached to specific label instances
+   labels: HashMap<Box<dyn SystemLabel>, SystemConfig>,
+   // Used to quickly look up system ids by the type ids of their underlying function
+   // when checking atomic ordering constraints
+   system_function_type_map: HashMap<TypeId, Vec<SystemId>>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: RawSystem,
+   config: SystemConfig,
+   // All of the ordering constraints from `SystemConfig` 
+   // are converted into individual systems.
+   // Only the systems that this system are after are stored here.
+   // This is what is actually checked by the scheduler to see if this system can be run.
+   // All if-needed orderings are either converted to strict orderings at this step or removed.
+   cached_ordering: CachedOrdering, 
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct RawSystem {
+   function: F,
+   // Contains all metadata and state
+   meta: SystemMeta
+   last_change_tick: u32,
+}
+```
+
+System configuration stores ordering dependencies, run criteria and labels.
+It can be created raw, or stored in association with either system labels or systems.
+
+```rust
+enum OrderingTarget {
+   Label(Box<dyn SystemLabel>),
+   System(SystemId),
+}
+
+struct SystemConfig {
+   strict_ordering_before: Vec<OrderingTarget>,
+   if_needed_ordering_before: Vec<OrderingTarget>,
+   atomic_ordering_before: Vec<OrderingTarget>,
+   strict_ordering_after: Vec<OrderingTarget>,
+   if_needed_ordering_after: Vec<OrderingTarget>,
+   atomic_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // This vector is always empty when this struct is used for labels
+   labels: Vec<Box dyn SystemLabel>,
+}
+```
+
+### Atomic ordering constraints

```

Comment:
Are atomic ordering constraints transitive?

Reply:
I think it'll make more sense if atomic constraints are defined as subgraphs rather than edges.